proposals/trampoline.md

```diff
@@ -0,0 +1,515 @@
++--------------------------------+
+|                                |
+|       I1 -----> I2 ------+     |
+|                          |     |
+|                          v     |
+|    I3 -----> I4 -----> Carol   |
+|                                |
++--------------------------------+
+```
+
+When Alice wants to pay that invoice, she chooses a trampoline node in her own
+neighborhood and simply provides the blinded paths in the trampoline onion. The
+trampoline node then forwards the payment using those blinded paths, which ensures
+that it doesn't learn the receiver's identity.
+
+```text
+        +---> N1 ---+                   +---> I4 ---+
+        |           |                   |           |
+Alice --+           +--> T1 -----> I3 --+           +--> Carol
+
+<------------------------><-------------------------------->
+     (normal payment)              (blinded payment)
+```
+
+The `update_add_htlc` message sent by Alice to N1 looks like:
+
+```text
+                        update_add_htlc
++---------------------------------------------------------------------------+
+| channel_id: 1x2x3                                                         |
+| htlc_id: 42                                                               |
+| amount_msat: 55000                                                        |
+| cltv_expiry: 1500                                                         |
+| payment_hash: 0xabcd                                                      |
+| onion_routing_packet:                                                     |
+| +-----------------------------------------------------------------------+ |
+| | +--------------------------------------------+----------------------+ | |
+| | | amt_to_forward: 54000                      | normal onion payload | | |
+| | | outgoing_cltv_value: 1450                  | for N1               | | |
+| | | short_channel_id: 5x3x7                    |                      | | |
+| | +--------------------------------------------+----------------------+ | |
+| | +--------------------------------------------+----------------------+ | |
+| | | amt_to_forward: 53000                      |                      | | |
+| | | outgoing_cltv_value: 1400                  | normal onion payload | | |
+| | | payment_secret: 0x1111                     | for T1               | | |
+| | | total_amount: 53000                        |                      | | |
+| | | trampoline_onion:                          |                      | | |
+| | | +---------------------------+------------+ |                      | | |
+| | | | amt_to_forward: 50000     |            | |                      | | |
+| | | | outgoing_cltv_value: 1000 | trampoline | |                      | | |
+| | | | blinded_paths:            | payload    | |                      | | |
+| | | |   <blinded_path_from_I1>  | for N1     | |                      | | |
+| | | |   <blinded_path_from_I3>  |            | |                      | | |
```

Comment:
@TheBlueMatt pointed out that just including the blinded paths for the trampoline hop this way currently won't work for sender-set custom TLVs or keysend preimages for async payments. Would be great to get your thoughts on this @t-bast!

Reply:
Good point! For this use-case, we'd need the recipient to support trampoline. This way we could include a trampoline onion for the recipient that would contain any sender-set data necessary. In this specific example, instead of having `EOF` right below, the trampoline onion would contain another trampoline hop for the recipient, that must be included by the trampoline node. Does that make sense?

Reply:
No, it's just like normal trampoline relay, intermediate nodes don't need to know anything about trampoline, only the final hop will receive a trampoline onion TLV in its onion payload.

Reply:
I'll need to think about it a bit more to figure out what's the best way of doing this indeed.

Reply:
Actually, thinking about this more, this shouldn't be too hard. When the recipient advertises support for trampoline in their invoice features, the sender should include a trampoline onion for them. This trampoline onion doesn't need to contain the `cltv_expiry`, it can contain only the amount that the sender intends to send, and extra TLVs (e.g. keysend). Then the intermediate trampoline node can use whatever block height when relaying that trampoline payment, and the recipient decides whether the CLTV expiry they receive is acceptable or not.

Reply:
Exactly, I'll update this PR with that scenario to ensure we're on the same page.

Reply:
Any status update on this, @t-bast? :)

Reply:
No, I didn't have time to work on this recently, and won't have time to look at it in the next couple of weeks, but let's discuss it again in two weeks!

Reply:
I thought about this more, and there's a trade-off I hadn't seen before, so I'd like your feedback on two potential designs. In https://github.com/lightning/bolts/commit/7589dd8e896eb2e3d206a05abb8a8c654057b693, I detailed how a sender can include blinded paths in the *trampoline* onion, which tells the trampoline node to relay the payment to those blinded paths. When done this way, the trampoline node is the *final* hop in the *trampoline* onion, so there's no way for the sender to include data for the recipient.

What follows is a bit subtle, so please take your time to check what is in the *trampoline* onion and what is in the *outer* onion in the two options below, and how it affects the payment.

### Option 1: simply add a trampoline hop for the recipient

My first idea was to slightly change the trampoline-to-blinded case from https://github.com/lightning/bolts/commit/7589dd8e896eb2e3d206a05abb8a8c654057b693 in the following way:

- the sender creates a *trampoline* packet where:
  - the last payload is for the final recipient (encrypted using its *blinded* node_id) and contains the amount sent and any additional TLV that we'd want to include (e.g. keysend)
  - the next-to-last payload contains the invoice's blinded paths
- the trampoline node decrypting the next-to-last payload:
  - detects that it contains blinded paths instead of an outgoing `node_id`, which tells it to relay to those blinded paths
  - must include in the *outer* onion of the outgoing payment the peeled trampoline packet (which contains the last payload for the final recipient)

The issue with that option is that the trampoline packet will be big, because it must contain blinded paths from the invoice. This isn't an issue between the sender and the last trampoline node, but it may be an issue for the trampoline node relaying to the blinded paths, because its *outer* onion will contain:

- a (big) trampoline onion in the last payload
- and all the `encrypted_recipient_data` necessary for standard blinded path logic (which are somewhat duplicated, because they used a lot of space in the trampoline onion, and now *also* appear in the outer onion)

The sender thus must limit how many blinded paths they include in the *trampoline* onion to ensure that the trampoline node will have enough space available to pay to those blinded paths. I'm afraid that in practice this will make it hard for the trampoline node to relay to those blinded paths because senders won't be able to include more than 1 or 2 blinded paths.

### Option 2: move blinded paths in the *outer* onion

The obvious workaround to this is that instead of putting the blinded paths in the *trampoline* onion, we can put them in the *outer* onion. This way the trampoline onion will be very small.

- the sender creates a *trampoline* packet where:
  - the last payload is for the final recipient (encrypted using its *blinded* node_id) and contains the amount sent and any additional TLV that we'd want to include (e.g. keysend)
  - the next-to-last payload does *not* contain blinded paths nor an outgoing `node_id`: it only contains the outgoing amount and cltv that the trampoline node must use
- the sender creates an *outer* packet where:
  - the last payload is for the trampoline node
  - the last payload contains the *trampoline* onion
  - the last payload contains the blinded paths
- the trampoline node decrypting the *outer* packet and its inner *trampoline* packet:
  - detects that it contains blinded paths and no outgoing `node_id`, which tells it to relay to those blinded paths
  - must include in the *outer* onion of the outgoing payment the peeled trampoline packet (which contains the last payload for the final recipient)

Since the trampoline onion can now be quite small, the trampoline node should have no issue relaying to the included blinded paths. The blinded paths only use space in the *outer* onion between the sender and the trampoline node. But the issue with this option is that the sender cannot use multiple trampoline nodes, since the blinded paths aren't in the *trampoline* onion but rather in the *outer* onion. Would that be an issue? Or is it fine because when combining trampoline and blinded paths, there is no need to use multiple trampoline hops?

If we think this is an issue, we could find a different way for the sender to encrypt some data for the recipient: but that would introduce yet another mechanism, so I'd like to avoid it to keep things simple.

Thoughts @valentinewallace @TheBlueMatt ?

Reply:
I think I see what you mean. To be clear, my understanding is that the trampoline packet would be too large in option 1 because even after the final trampoline hop unwraps its layer, per standard onion forwarding it essentially "pads" the trampoline onion that it forwards to be the same length as the one that it received. It's not completely obvious to me that there can't be a way around that, but I haven't looked into it. 

In any case, IMO option 2 doesn't sound unreasonable and sounds simpler. Would like to get @TheBlueMatt's thoughts on this though.

Reply:
> the trampoline packet would be too large in option 1 because even after the final trampoline hop unwraps its layer, per standard onion forwarding it essentially "pads" the trampoline onion that it forwards to be the same length as the one that it received

Yes, that's exactly it!

> It's not completely obvious to me that there can't be a way around that, but I haven't looked into it.

We cannot work around that unless we stop doing standard Sphinx, and thus lose its privacy properties...if we think we should work around it, we should find a completely different way of encrypting data only for the recipient, but I'd like to avoid that to avoid creating too many new protocols.

> In any case, IMO option 2 doesn't sound unreasonable and sounds simpler. Would like to get @TheBlueMatt's thoughts on this though.

I think it's reasonable too: having multiple trampoline hops doesn't give you any additional privacy when using blinded paths, and only has drawbacks (longer paths so more likely to fail). So it seems fine to me to use a single trampoline node when paying a trampoline recipient using blinded paths.

Reply:
I don't think that attack would work: the recipient sees that the `invoice_request` contains the trampoline feature bit and returns an invoice that contains the trampoline feature bit as `mandatory` (since the payer told them they support it). The recipient can then just fail the payment if it doesn't contain a trampoline onion.

We must make sure the trampoline node cannot create the trampoline onion themselves though, but that's easy to do by using the `payer_key` somehow (just including it again in the trampoline payload is probably enough?).

For the case where we're using trampoline to pay a blinded path where the recipient does *not* support trampoline, the trampoline node is not able to steal some sats as long as the payer pays exactly the invoice amount (which they decide in the `invoice_request`, so it should be ok?).

Reply:
Still thinking on this, but note that we may not be able to require trampoline feature support in invoices post-#1149, since async recipients will be creating static invoices in advance and will want to accept payment from non-trampoline-supporting senders.

Reply:
Right, but in the case where the recipient cannot come online to validate the payment, it knows that there is some trust involved anyway?

I'm still not convinced that it is a good idea, I'd rather always wait for the recipient to come online and generate a real invoice (and otherwise fail the payment), even with all the caveats it creates when waking up the payer or recipient takes some time...I'm not sure we'll ever implement the static invoice part in eclair, but if you want it for LDK I'm fine with having it in the spec, it's your call!

Reply:
> Sure, but there's nothing specific about the sender's trampoline onion? The malicious intermediary can also create a new trampoline onion.

Right, that's what was addressed in the next paragraph, where we need to reuse the `payer_key` to tie the `invoice_request` to the payment. But if you feel this is a layering violation, we'll need to come up with something else...

> If we can provide the trampoline with a list of (intro point, onion tail) that'd be much better than revealing the blinded path itself.

I'm not sure how we could do something like this. It seems to me exactly like rendezvous routing, which lacks flexibility for lightning payments because it doesn't let the trampoline node:

- set the expiry themselves (which is particularly important for async payments)
- set the amount themselves if they need to do MPP

I'd be happy to see a better proposal than my current one though if you find one, but the devil really is in the details, so it needs to be quite detailed to see if it would work or not.

Reply:
> Okay, Val pointed out that (duh) if an intermediary replaces the onion for an async payment the recipient can't claim the payment cause they won't have the preimage.

Right, I don't think the trampoline node is able to cheat if:

- we tie the trampoline onion to the sender's `payer_id` _somehow_ (an ECDH like you mention is probably the simplest way to do it without wasting additional space in the onion), which guarantees that if a trampoline onion is provided, it was created by the node who sent `invoice_request` (and thus received the invoice)
- if the recipient created an invoice with the trampoline feature bit (because the `invoice_request` advertised that feature bit as well), it rejects payments that don't contain a trampoline onion, which guarantees that the trampoline node has to include the trampoline onion
- if the recipient did *not* create the invoice themselves (which matches the async payment case when the recipient cannot be woken up to receive `invoice_request`), then we need some form of keysend anyway, which will only work if the trampoline node correctly includes the sender's trampoline onion

I can specify that in more details and add it to #836 and to this PR. I'll also need to decide between options 1 and 2 from [my previous comment](https://github.com/lightning/bolts/pull/829#discussion_r1679382517). I think we should do option 2), which restricts those payments to using a single trampoline node (see details in the comment). Are you ok with that?

Reply:
> ...  I'll also need to decide between options 1 and 2 from [my previous comment](https://github.com/lightning/bolts/pull/829#discussion_r1679382517). I think we should do option 2), which restricts those payments to using a single trampoline node (see details in the comment). Are you ok with that?

I think the trade-off of a single trampoline node seems reasonable if it means that the sender can pass more blinded paths from the receiver to their trampoline node. Also if the sender pre-computes how many blinded paths they can include differently than their trampoline node we'll have some weird interop bugs. 

As I was thinking about these options, I [created some diagrams to rift on your examples](https://gist.github.com/remyers/3d31b86107dea20d338c9e4acbe1e9f3) from #836 to depict the two options. Perhaps others will find it useful.

Reply:
Could the sender add hops to the recipient's blinded paths to increase privacy, using `next_blinding_override`?

Reply:
IIRC another possibility we touched on was including trampoline hops within the blinded path. In my mind, this essentially meant that a forwarding trampoline+blinded node's TLVs would contain a `next_node_id` rather than an scid (amongst other changes). @t-bast does this option still seem worth exploring too? Happy to help btw.